src/coreclr/jit/valuenum.cpp

```diff
@@ -11161,16 +11161,55 @@ void Compiler::fgValueNumberPhiDef(GenTreeLclVar* newSsaDef, BasicBlock* blk, bo
 
         ValueNumPair phiArgVNP = lvaGetDesc(phiArg)->GetPerSsaData(phiArg->GetSsaNum())->m_vnPair;
 
-#ifdef DEBUG
-        if (verbose && isUpdate && (phiArgVNP != phiArg->gtVNPair))
+        if (isUpdate && (phiArgVNP != phiArg->gtVNPair))
         {
-            printf("Updating phi arg [%06u] VN from ", dspTreeID(phiArg));
-            vnpPrint(phiArg->gtVNPair, 0);
-            printf(" to ");
-            vnpPrint(phiArgVNP, 0);
-            printf("\n");
-        }
+            bool canUseNewVN = false;
+
+            // We can potentially refine this phi arg.
+            // Make sure the new phi arg VN is loop invariant.
+            //
+            FlowGraphNaturalLoop* const vnLoop = vnStore->LoopOfVN(phiArgVNP.GetConservative());
```

Comment:
Is this sufficient? Doesn't it need to use the same logic as `optVNIsLoopInvariant`?

Reply:
E.g. for
```csharp
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Problem(int x)
    {
        int y = 0;
        while (x != 0)
        {
            if (y == x + 1) return -1;
            y = x + 1;
            Update(ref x);
        }
        return x;
    }
```
the VN will be `VNF_ADD` and `LoopOfVN` will return nullptr

Reply:
Ah, right.

We can directly use `optVNIsLoopInvariant` since we only ever try updating phis for loops.